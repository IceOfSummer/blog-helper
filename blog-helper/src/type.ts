
/**
 * html 富文本字符串
 */
type HTML = string

export interface MarkdownParser {
  /**
   * 解析markdown
   */
  parse(markdown: string): Promise<HTML>
}

export type WebVisitPath = string[]


export type BaseMetadata = {
  /**
   * 是否为首页
   */
  isHomePage?: boolean,
}

export type DatasourceItem<UserData extends BaseMetadata> = {
  /**
   * 唯一标识符, 通常是文件访问路径
   */
  id: string;
  /**
   * 访问路径
   */
  visitPath: WebVisitPath
  /**
   * 文件类型，可以是文件路径，或者文件类型拓展符
   */
  type: string
  /**
   * 文件路径
   */
  filepath: string
  /**
   * 用户通过回调函数添加的元数据
   */
  metadata: UserData
}

