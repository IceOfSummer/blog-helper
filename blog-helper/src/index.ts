export { default as searchPages } from './datasource/search'
export { splitMarkdownContent } from './spliter'
export { default as adjustTocPlugin }  from './parser/toc-adjust-plugin'
export { default as generateHeadingIdPlugin } from './parser/generate-heading-id-plugin'
export { default as createPageWithIndexBuilder } from './holder/page-index-helper'