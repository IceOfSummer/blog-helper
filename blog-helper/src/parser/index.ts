import { unified } from 'unified'
import remarkParse from 'remark-parse'
import remarkGfm from 'remark-gfm'
import adjustToc from './toc-adjust-plugin'
import remarkRehype from 'remark-rehype'
import rehypeStringify from 'rehype-stringify'
import generateHeadingId from "./generate-heading-id-plugin";


export function createMdParser() {

  const markdownToHtml = async (markdownContent: string): Promise<string> => {
    const parsed = await unified()
      .use(remarkParse)
      .use(remarkGfm)
      .use(adjustToc)
      .use(remarkRehype, { allowDangerousHtml: true })
      .use(generateHeadingId)
      .use(rehypeStringify, { allowDangerousHtml: true })
      .process(markdownContent)
    return String(parsed)
  }

  return {
    async parse(markdown: string): Promise<string> {
      return markdownToHtml(markdown)
    }
  }
}

