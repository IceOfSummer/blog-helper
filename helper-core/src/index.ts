export { FileSystemDatasource } from './datasource/FileSystemDatasource'
export { splitMarkdownContent } from './parser'
