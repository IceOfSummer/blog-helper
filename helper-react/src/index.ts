export { default as createMdParser } from './impl/md/MdParser'
export { default as createMdxParser } from './impl/mdx/MdxParser'