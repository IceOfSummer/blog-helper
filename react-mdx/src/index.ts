export { default as createMdxParser } from './MdxParser'