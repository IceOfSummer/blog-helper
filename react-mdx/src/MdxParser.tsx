import React, {ReactNode} from 'react'
import { compile, run } from '@mdx-js/mdx'
import * as runtime from 'react/jsx-runtime'
import * as os from "node:os";
import createCommonHeadingWithId from './CommonHeadingWithId'
import {adjustTocPlugin} from 'blog-helper'
import remarkGfm from 'remark-gfm'
import type {MDXProps} from "mdx/types";


type ParseError = Error & {
  line: number
  name: string
  column: number
  place: {
    line: number
    column: number
    offset: number
  }
}

function isParseAeeror(err: unknown): err is ParseError {
  if (!err || !(typeof err === 'object')) {
    return false
  }
  const errObj = err as Partial<ParseError>
  return !!errObj.column && !!errObj.line && !!errObj.name && !!errObj.place
}

function takeAboveLines(self: string[], lines: string[], startLine: number, lineCnt: number) {
  const cnt = Math.min(startLine, lineCnt - 1)

  for (let i = startLine - cnt; i <= startLine; i++) {
    self.push(` ${i.toString(10).padStart(4, ' ')} | ${lines[i]}`)
  }
}

export type MdxParserConfig = {
  components: MDXProps['components']
}

function createMdxParser(config?: MdxParserConfig) {

  const components = {
    h1: createCommonHeadingWithId('h1'),
    h2: createCommonHeadingWithId('h2'),
    h3: createCommonHeadingWithId('h3'),
    h4: createCommonHeadingWithId('h4'),
    h5: createCommonHeadingWithId('h5'),
    ...config?.components,
  }


  async function parseMarkdownContent0(content: string): Promise<ReactNode> {
    const code = String(await compile(content, { outputFormat: 'function-body', remarkPlugins: [adjustTocPlugin, remarkGfm] }))

    const { default: MDXContent } = await run(code, {
      ...runtime,
      // @ts-ignore
      baseUrl: import.meta.url,
    })

    // Render the MDX content, supplying the ClientComponent as a component
    return <MDXContent components={components} />
  }

  /**
   * 处理 html 博客内容
   */
  async function parseMarkdownContent(content: string): Promise<React.ReactNode> {
    try {
      return await parseMarkdownContent0(content)
    } catch (e) {
      if (!isParseAeeror(e)) {
        return Promise.reject(e)
      }
      // indicate where is wrong.
      const lines = content.split(os.EOL)
      const message = ['Failed to parse markdown content:']
      const programLine = e.line - 1

      message.push('')
      takeAboveLines(message, lines, programLine, 4)

      message.push('        ' + (new Array(e.column - 1).join(' ')) + '^')
      message.push('\t' + e.message)


      return Promise.reject(new Error(message.join(os.EOL)))
    }
  }


  return  {
    async parse(markdown: string): Promise<React.ReactNode> {

      return await parseMarkdownContent(markdown)
    }
  }
}

export default createMdxParser
