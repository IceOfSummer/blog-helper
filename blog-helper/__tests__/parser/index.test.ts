import { expect, test } from '@jest/globals'
import { createMdParser } from '../../src/parser'
import fs from 'node:fs'


test('Test markdown render', async () => {
  const parser = createMdParser()
  const result = await parser.parse(fs.readFileSync('__tests__/__source__/markdown/hello.md', { encoding: 'utf-8' }))
  expect(result).toMatchSnapshot()
})