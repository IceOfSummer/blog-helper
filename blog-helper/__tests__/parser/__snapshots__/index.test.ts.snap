// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test markdown render 1`] = `
"<hr>
<h2 id="title:-example
date:-2024/07/31-16:22:39">title: example
date: 2024/07/31 16:22:39</h2>
<p>Markdown是一种轻量级的「标记语言」。</p>
<h2 id="1-Markdown语法教程">1 Markdown语法教程</h2>
<h3 id="1.1-标题">1.1 标题</h3>
<p>不同数量的<code>#</code>可以完成不同的标题.。</p>
<h3 id="1.2-字体">1.2 字体</h3>
<p>粗体、斜体、粗体和斜体，删除线，需要在文字前后加不同的标记符号。如下：</p>
<p><strong>这个是粗体</strong></p>
<p><em>这个是斜体</em></p>
<p><em><strong>这个是粗体加斜体</strong></em></p>
<p><del>这里是删除线</del></p>
<p>注：如果想给字体换颜色、字体或者居中显示，需要使用内嵌HTML来实现。</p>
<h3 id="1.3-无序列表">1.3 无序列表</h3>
<p>无序列表的使用，在符号<code>-</code>后加空格使用。如下：</p>
<ul>
<li>无序列表 1</li>
<li>无序列表 2</li>
<li>无序列表 3</li>
</ul>
<p>如果要控制列表的层级，则需要在符号<code>-</code>前使用空格。如下：</p>
<ul>
<li>无序列表 1</li>
<li>无序列表 2
<ul>
<li>无序列表 2.1</li>
<li>无序列表 2.2</li>
</ul>
</li>
</ul>
<h3 id="1.4-有序列表">1.4 有序列表</h3>
<p>有序列表的使用，在数字及符号<code>.</code>后加空格后输入内容，如下：</p>
<ol>
<li>有序列表 1</li>
<li>有序列表 2</li>
<li>有序列表 3</li>
</ol>
<h3 id="1.5-引用">1.5 引用</h3>
<p>引用的格式是在符号<code>></code>后面书写文字。如下：</p>
<blockquote>
<p>读一本好书，就是在和高尚的人谈话。 ——歌德</p>
</blockquote>
<blockquote>
<p>雇用制度对工人不利，但工人根本无力摆脱这个制度。 ——阮一峰</p>
</blockquote>
<h3 id="1.6-链接">1.6 链接</h3>
<p><a href="https://github.com/nextjs-particlex-theme/particlex">particlex</a></p>
<h3 id="1.7-图片">1.7 图片</h3>
<p><img src="/favicon.ico" alt="favicon.ico"></p>
<h3 id="1.8-分割线">1.8 分割线</h3>
<p>可以在一行中用三个以上的减号来建立一个分隔线，同时需要在分隔线的上面空一行。如下：</p>
<hr>
<h3 id="1.9-表格">1.9 表格</h3>
<p>可以使用冒号来定义表格的对齐方式，如下：</p>
<table>
<thead>
<tr>
<th align="left">姓名</th>
<th align="center">年龄</th>
<th align="right">工作</th>
</tr>
</thead>
<tbody>
<tr>
<td align="left">小可爱</td>
<td align="center">18</td>
<td align="right">吃可爱多</td>
</tr>
<tr>
<td align="left">小小勇敢</td>
<td align="center">20</td>
<td align="right">爬棵勇敢树</td>
</tr>
<tr>
<td align="left">小小小机智</td>
<td align="center">22</td>
<td align="right">看一本机智书</td>
</tr>
</tbody>
</table>
<h3 id="1.10-代码块">1.10 代码块</h3>
<p>如果在一个行内需要引用代码，只要用反引号引起来就好，如下：</p>
<p>Use the <code>printf()</code> function.</p>
<p>在需要高亮的代码块的前一行及后一行使用三个反引号，同时<strong>第一行反引号后面表示代码块所使用的语言</strong>，如下：</p>
<pre><code class="language-java">// FileName: HelloWorld.java
public class HelloWorld {
  // Java 入口程序，程序从此入口
  public static void main(String[] args) {
    System.out.println("Hello,World!"); // 向控制台打印一条语句
  }
}
</code></pre>
<pre><code class="language-xml">&#x3C;hello>
	&#x3C;world>&#x3C;/world>
&#x3C;/hello>
</code></pre>
<h2 id="2-其他语法">2 其他语法</h2>
<h3 id="2.1-HTML">2.1 HTML</h3>
<p>支持原生 HTML 语法，请写内联样式，如下：</p>
<p><span style="display:block;text-align:right;color:orangered;">橙色居右</span>
<span style="display:block;text-align:center;color:orangered;">橙色居中</span></p>"
`;
